src/uu/cksum/src/cksum.rs

```diff
@@ -5,109 +5,140 @@
+use sha1::Sha1;
+use sha2::{Sha224, Sha256, Sha384, Sha512};
+use std::ffi::OsStr;
 use std::fs::File;
 use std::io::{self, stdin, BufReader, Read};
+use std::iter;
 use std::path::Path;
-use uucore::display::Quotable;
-use uucore::error::{FromIo, UResult};
-use uucore::{format_usage, show};
-
-// NOTE: CRC_TABLE_LEN *must* be <= 256 as we cast 0..CRC_TABLE_LEN to u8
-const CRC_TABLE_LEN: usize = 256;
-const CRC_TABLE: [u32; CRC_TABLE_LEN] = generate_crc_table();
+use uucore::{
+    error::{FromIo, UResult},
+    format_usage,
+    sum::{div_ceil, Digest, DigestWriter, BSD, CRC, SYSV},
+};
 
 const USAGE: &str = "{} [OPTIONS] [FILE]...";
 const ABOUT: &str = "Print CRC and size for each file";
 
-const fn generate_crc_table() -> [u32; CRC_TABLE_LEN] {
-    let mut table = [0; CRC_TABLE_LEN];
-
-    let mut i = 0;
-    while i < CRC_TABLE_LEN {
-        table[i] = crc_entry(i as u8);
-
-        i += 1;
-    }
-
-    table
-}
-
-const fn crc_entry(input: u8) -> u32 {
-    let mut crc = (input as u32) << 24;
-
-    let mut i = 0;
-    while i < 8 {
-        let if_condition = crc & 0x8000_0000;
-        let if_body = (crc << 1) ^ 0x04c1_1db7;
-        let else_body = crc << 1;
-
-        // NOTE: i feel like this is easier to understand than emulating an if statement in bitwise
-        //       ops
-        let condition_table = [else_body, if_body];
-
-        crc = condition_table[(if_condition != 0) as usize];
-        i += 1;
+fn detect_algo(program: &str) -> (&'static str, Box<dyn Digest + 'static>, usize) {
+    match program {
+        "sysv" => ("SYSV", Box::new(SYSV::new()) as Box<dyn Digest>, 512),
+        "bsd" => ("BSD", Box::new(BSD::new()) as Box<dyn Digest>, 1024),
+        "crc" => ("CRC", Box::new(CRC::new()) as Box<dyn Digest>, 256),
+        "md5" => ("MD5", Box::new(Md5::new()) as Box<dyn Digest>, 128),
+        "sha1" => ("SHA1", Box::new(Sha1::new()) as Box<dyn Digest>, 160),
+        "sha224" => ("SHA224", Box::new(Sha224::new()) as Box<dyn Digest>, 224),
+        "sha256" => ("SHA256", Box::new(Sha256::new()) as Box<dyn Digest>, 256),
+        "sha384" => ("SHA384", Box::new(Sha384::new()) as Box<dyn Digest>, 384),
+        "sha512" => ("SHA512", Box::new(Sha512::new()) as Box<dyn Digest>, 512),
+        "blake2b" => (
+            "BLAKE2",
+            Box::new(blake2b_simd::State::new()) as Box<dyn Digest>,
+            512,
+        ),
+        "sm3" => ("SM3", Box::new(sm3::Sm3::new()) as Box<dyn Digest>, 512),
+        _ => panic!("unknown algorithm"),
     }
-
-    crc
 }
 
-#[inline]
-fn crc_update(crc: u32, input: u8) -> u32 {
-    (crc << 8) ^ CRC_TABLE[((crc >> 24) as usize ^ input as usize) & 0xFF]
+struct Options {
+    algo_name: String,
+    digest: Box<dyn Digest + 'static>,
+    output_bits: usize,
 }
 
-#[inline]
-fn crc_final(mut crc: u32, mut length: usize) -> u32 {
-    while length != 0 {
-        crc = crc_update(crc, length as u8);
-        length >>= 8;
+#[allow(clippy::cognitive_complexity)]
```

Comment:
could please document the function profile ?

Reply:
not sure how to fix this one

Reply:
https://doc.rust-lang.org/rust-by-example/meta/doc.html  :)

Reply:
how about this?